### ISSUE TITLE
explain `insn/api: x64 nested thunk` test case

### ISSUE THREAD

**Issue by williballenthin** [fireeye/capa#356]
[here](https://github.com/fireeye/capa/blob/0e009c7c1251fe69a95cfbf6da1199378c88d733/tests/fixtures.py#L380) is a unit test that we use to assert that our analysis backends consistently handle chains of thunks:

```py
# insn/api: x64 nested thunk
("82bf6", 
 "function=0x140059342", 
 capa.features.insn.API("ElfClearEventLogFile"), 
 True),
```

here's how we currently understand "thunks" within capa:

> within capa, the term "thunk" only really refers to "import thunk" which would be a chain of unconditional jmps *to an imported API*. usually you'd see `call -> jmp -> API`. in the test case, we see `call -> jmp -> jmp -> API`. in these cases, we want a call to the thunk to be recorded simply as a call to the imported API.
>
> if there is anything other than an imported API function at the destination of the jump, then capa shouldn't handle the case specially.

turning our attention to the test, here's what we see:

### 0x140059342

![image](https://user-images.githubusercontent.com/156560/97729197-11becc00-1a98-11eb-82e0-ebf5074acd85.png)

0x140059342 is a function that immediately tail calls to function 0x14005E0C0. you could call this a "thunk" too, though this is not supposed to be a meaningful part of the test. more on this later.

### 0x14005E0C0
![image](https://user-images.githubusercontent.com/156560/97729286-3024c780-1a98-11eb-8d3b-77971c7d6af0.png)

0x14005E0C0 calls `j_ElfClearEventLogFileW` a few times. here's the meat of the test:

`j_ElfClearEventLogFileW` -jmp> `ElfClearEventLogFileW` -jmp> `__imp_ElfClearEventLogFileW`

calls to `j_ElfClearEventLogFileW` should be extracted API features for `ElfClearEventLogFileW` even though there are two chained unconditional jumps.

![image](https://user-images.githubusercontent.com/156560/97729317-39ae2f80-1a98-11eb-83f4-7272310cd501.png)

![image](https://user-images.githubusercontent.com/156560/97729355-43379780-1a98-11eb-8d4f-077602b2e3a6.png)

![image](https://user-images.githubusercontent.com/156560/97729384-4b8fd280-1a98-11eb-95b5-62f8badf8344.png)

## sidebar: tail calls
a source of confusion within this test is how to decide what a function is. its up to analysis backends to decide what a function is, and i'm not sure if/how capa will enforce this. probably, we should do our best to work with all backends.

here's what i see:

0x140059342 could be considered a "thunk" (and/or tail call) since its whole body is to unconditionally jump to 0x14005E0C0 that seems to be its own function.

0x140059342 could be considered its own function, since its the target of a call: 
![image](https://user-images.githubusercontent.com/156560/97731391-a7f3f180-1a9a-11eb-9734-429f7ed297b5.png)

0x14005E0C0 could be considered its own function, since its registered in the `RUNTIME_FUNCTION` exception handler array.
![image](https://user-images.githubusercontent.com/156560/97731419-b3471d00-1a9a-11eb-9af4-1796da67d935.png)

if both of the above are true, then the two functions share a lot of instructions and their features. there's nothing wrong with this, and sharing basic blocks/instructions should be supported by any modern analysis tool. viv falls into this category, i believe.

if one of the above is not true, that's fine too. at least one should probably be a function, though. IDA falls into this category.

cc: @danielplohmann who initially brought up this question.

**Comment by williballenthin** [comment]
@mike-hunhoff any additional insight here?

**Comment by mike-hunhoff** [comment]
Thank you for bringing up this question @danielplohmann.

@williballenthin you discussed this quite thoroughly already so not much more to add. Here, as you mentioned, the result is the same in terms of generated features, regardless of how viv and IDA define the function boundary, but these differences happen to result in a failed unit test. One option would be to simply compensate for differences like this in our unit tests. Another option would be to enforce some level of expectation for analysis results produced by a backend e.g. how function boundaries are established. I do not believe the latter option is preferred simply due to the level of complexity this may introduce to backend development and I agree that we should do are best to work with all backends.

**Comment by danielplohmann** [comment]
Hi all, 

as a possible solution, I would propose to find an equivalent code example for the functionality that the test intends to cover, i.e. a x64 nested thunk. I quickly scouted all of your test files for such cases and located some in the following binaries.
 * `82BF6347ACF15E5D883715DC289D8A2B.exe_` - 10 cases, all with same characteristics (tailcalls) as described by @williballenthin above.
 * `al-khaser_x64.exe_` - 122 cases of nested thunks but actually some being used in functions that are not tailcalled. For instance `sub_14004B4F0` makes use of nested thunks (like `0x14004B532 call j___vcrt_GetModuleHandleW`) but is called in regular fashion by `sub_14004BB40`. 
I did not check yet if this function is recognized properly by vivisect, but if yes, it might be a good candidate for a replacement. The downside is that `al-khaser_x64.exe_` is currently not used in other tests and thus would mean more runtime for the fixtures overall.

I did not locate other cases of nested thunks (at least for x64) in your collection of test files.
